shlr/gdb/src/packet.c

```diff
@@ -1,157 +1,181 @@
 /* libgdbr - LGPL - Copyright 2014-2016 - defragger */
 
 #include "packet.h"
+#include "utils.h"
 
-#define DELTA_CHECK 0
+#define READ_TIMEOUT (250 * 1000)
 
-char get_next_token(parsing_object_t* po) {
-	return po->buffer[po->position++];
-}
+enum {
+	HEADER	= 1 << 0,
+	CHKSUM	= 1 << 1,
+	DUP	= 1 << 2,
+	ESC	= 1 << 3,
+};
 
-void handle_escape(parsing_object_t* po) {
-	char token;
-	if (po->position >= po->length) return;
-	token = get_next_token (po);
-	if (token == '}') handle_data (po);
-	else handle_escape (po);
-}
+struct parse_ctx {
+	unsigned long flags;
+	unsigned char last;
+	unsigned char chksum;
+	int chksum_nibble;
+};
 
-void handle_chk(parsing_object_t* po) {
-	char checksum[3];
-	if (po->position >= po->length) return;
-	checksum[0] = get_next_token (po);
-	checksum[1] = get_next_token (po);
-	checksum[2] = '\0';
-	po->checksum = (uint8_t) strtol (checksum, NULL, 16);
-}
+static int append(libgdbr_t *g, char ch)
+{
+	char *ptr;
 
-void handle_data(parsing_object_t* po) {
-	char token;
-	if (po->position >= po->length) return;
-	token = get_next_token (po);
-	if (token == '#') {
-		po->end = po->position - 1; // subtract 2 cause of # itself and the incremented position after getNextToken
-		handle_chk (po);
-	} else if (token == '{') {
-		handle_escape (po);
-	} else handle_data (po);
-}
+	if (g->data_len == g->data_max) {
+		ptr = realloc(g->data, g->data_max * 2);
+		if (!ptr) {
+			eprintf("%s: Failed to reallocate buffer\n", __func__);
+			return -1;
+		}
 
-void handle_packet(parsing_object_t* po) {
-	char token;
-	if (po->position >= po->length) return;
-	token = get_next_token (po);
-	if (token == '$') {
-		po->start = po->position;
-		handle_data (po);
-	}
-	else if	(token == '+') {
-		po->acks++;
-		handle_packet (po);
+		g->data = ptr;
+		g->data_max *= 2;
 	}
+
+	g->data[g->data_len++] = ch;
+	return 0;
 }
 
-/**
- * helper function that should unpack
- * run-length encoded streams of data
- */
-int unpack_data(char* dst, char* src, uint64_t len) {
+static int unpack(libgdbr_t *g, struct parse_ctx *ctx, int len)
+{
 	int i = 0;
-	char last = 0;
-	int ret_len = 0;
-	char* p_dst = dst;
-	while (i < len) {
-		if (src[i] == '*') {
-			if (++i >= len) {
-				eprintf ("Runlength decoding error\n");
+	int j = 0;
+
+	for (i = 0; i < len; i++) {
+		char cur = g->read_buff[i];
+
+		if (ctx->flags & CHKSUM) {
+			ctx->chksum -= hex2int(cur) << (ctx->chksum_nibble * 4);
+
+			if (!--ctx->chksum_nibble)
```

Comment:
If the 'if' statement is on two lines, like this case, you should wrap it with braces.